### PR TITLE
feat(harness): require role field in Validate()

### PR DESCRIPTION
**Pull request by ggallen**
## Summary

- Promotes the missing `role:` check from a `Lint()` warning to a `Validate()` hard error, as specified in ADR-0045 Phase 4 PR 1
- Every scaffold harness already sets `role:`, so this is non-breaking for existing users
- Removes the now-redundant lint diagnostic for missing role
- Updates all test fixtures across harness, CLI run, lock, and lock-all tests to include `role:` where needed

## Context

This is Phase 4, PR 1 of [ADR-0045 (Forge-Portable Harness Schema)](docs/ADRs/0045-forge-portable-harness-schema.md). The `role:` field was introduced in Phase 2 as optional with a lint warning. This PR makes it required at validation time, enforcing the contract for all harness configurations going forward.

## Test plan

- [x] `go test ./internal/harness/` — all harness validation, lint, forge, compose, and integration tests pass
- [x] `go test ./internal/cli/` — all run, lock, and lock-all tests pass
- [x] `go test ./internal/...` — full internal test suite passes
- [x] Pre-commit hooks pass (gofmt, go vet, etc.)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Review by ralphbean — APPROVED** [review]
LGTM.

File: internal/harness/harness.go [hunk 1]
```diff
@@ -198,7 +198,7 @@ type Harness struct {
 	Agent                  string                  `yaml:"agent"`
 	Doc                    string                  `yaml:"doc,omitempty"` // source-repo-only; not resolved at runtime, used by lint-agent-docs
```
Comment (waynesun09) [inline internal/harness/harness.go:199]: **[medium] design** — Removing omitempty from Role yaml tag causes empty string to serialize in YAML output

Changing the Role tag from `yaml:"role,omitempty"` to `yaml:"role"` means that when a Harness struct is serialized to YAML (e.g., during scaffold generation, lock file creation, or any code path that marshals a Harness), the `role:` key will always be emitted, even when Role is the zero value (empty string). This produces `role: ""` in output YAML, which is cosmetically undesirable and could confuse users reading generated harness files.

Since Validate() now enforces that Role must be non-empty, the omitempty removal is logically consistent (a valid harness always has a role), but any code path that marshals a partially-constructed Harness before validation will produce the empty field.

Suggestion: Verify that no code path marshals a Harness to YAML before Validate() has been called. If any such path exists (e.g., scaffold template generation, test helpers), the empty `role: ""` will appear in output. Consider keeping `omitempty` on the tag since the Validate() check is the real enforcement mechanism, or audit all marshal call sites to confirm they only emit fully-validated harnesses.

File: internal/harness/lint.go [hunk 2]
```diff
@@ -39,14 +39,5 @@ func (d Diagnostic) String() string {
...
-
-	if h.Role == "" {
-		diags = append(diags, Diagnostic{
-			Severity: SeverityWarning,
-			Field:    "role",
-			Message:  "role is not set; it will be required in a future version",
-		})
-	}
-
 	return diags
```
Comment (waynesun09) [inline internal/harness/lint.go]: **[medium] design** — Lint() is now an empty function that always returns nil

After removing the role-missing diagnostic, Lint() now has no diagnostics to check and always returns nil (line 42-44: `var diags []Diagnostic; return diags`). While this is correct for the current phase, it leaves a function that does nothing.

The ADR-0045 design decision D3 established Lint() as a separate method for non-fatal warnings. Keeping the empty function is reasonable as infrastructure for future diagnostics, but the function body could be simplified to `return nil` to make the intent clear. More importantly, the lint_test.go tests now only verify that Lint() returns nil in all cases, which provides no value as regression tests.

Suggestion: Consider simplifying the body to `return nil` for clarity. The test coverage for Lint() is now trivially passing in all cases -- the tests verify that an empty function returns nil, which is not useful. If future Lint() diagnostics are planned (e.g., slug-missing warning), document that intent in a comment. If no additional diagnostics are planned, consider whether Lint() should be removed entirely to avoid dead code.

File: internal/cli/run_test.go [hunk 3]
```diff
@@ -1821,45 +1820,6 @@ func TestRunAgent_LintWarningOnMissingRole(t *testing.T) {
 	repoDir := t.TempDir()
```
Comment (waynesun09) [inline internal/cli/run_test.go:1820]: **[medium] testing** — TestRunAgent_ErrorOnMissingRole does not verify the error comes from Validate vs. other sources

The renamed test `TestRunAgent_ErrorOnMissingRole` (line 1799) asserts that `err.Error()` contains `"role field is required"` but does not verify that the error originates from harness validation specifically. The test creates a harness file without a role field and calls runAgent, which goes through the full loading pipeline including potential LoadWithBase, forge resolution, etc.

The assertion `assert.Contains(t, err.Error(), "role field is required")` correctly verifies the error message, but if a future code change adds a different "role field is required" error message from another validation layer, this test could pass spuriously. Additionally, the test no longer verifies that the error occurs early (before reaching openshell), unlike the old test which explicitly checked the error was about openshell (meaning the lint warning was non-fatal and the pipeline continued).

Suggestion: Consider also asserting that the error message contains "invalid harness" (the wrapper added by Load/LoadWithOpts) to confirm the error comes from the Validate() call specifically, e.g.: `assert.Contains(t, err.Error(), "invalid harness: role field is required")`. This makes the test more precise about which code path produces the error.

File: internal/harness/harness_test.go [hunk 4]
```diff
@@ -1092,18 +1110,17 @@ slug: fullsend-ai-triage
 	assert.Equal(t, "fullsend-ai-triage", h.Slug)
 }
 
-func TestLoad_RoleAndSlugAbsent(t *testing.T) {
+func TestLoad_RoleMissing(t *testing.T) {
 	content := `
 agent: agents/test.md
```
Comment (waynesun09) [inline internal/harness/harness_test.go:1115]: **[medium] testing** — TestLoad_RoleMissing now tests the same thing as TestValidate_RoleValid negative cases

The updated `TestLoad_RoleMissing` test now expects a hard error with message "role field is required", which is effectively the same check as creating a Harness with empty Role and calling Validate() directly. The test goes through the full Load() path (file I/O + unmarshal + validate), but the actual validation logic being tested is identical to what `TestValidate_RoleInvalid` test cases cover.

However, the original `TestLoad_RoleMissing` test served a different purpose before this PR: it verified that a harness without role could still be loaded successfully (role was optional). The test has been updated to verify the new behavior, which is correct.

This is a low-impact observation -- the test duplication is minor and the Load()-level test does exercise the full pipeline. No action required.

File: internal/harness/harness.go [hunk 5]
```diff
@@ -198,7 +198,7 @@ type Harness struct {
...
 	Description            string                  `yaml:"description,omitempty"`
```
Comment (waynesun09) [inline internal/harness/harness.go:200]: **[medium] adr-alignment** — Role YAML tag removes omitempty but YAML marshaling will emit empty role as `role: ""`

The YAML tag changed from `yaml:"role,omitempty"` to `yaml:"role"`. This means when marshaling a Harness struct with an empty Role, YAML output will include `role: ""` instead of omitting the field entirely. This is correct for Phase 4 since role is now required and should always be present, but it has a subtle interaction: any code path that creates a Harness struct programmatically (e.g., for base composition in compose.go's mergeHarness, or in tests that construct Harness literals) and then marshals it will emit the empty `role:` key. If downstream tooling reads these marshaled outputs, it could encounter `role: ""` which would fail Validate(). Verify that no code path marshals a partially-constructed Harness before role is populated. This aligns with ADR-0045's intent but the marshaling side-effect should be documented or tested.

File: internal/harness/lint.go [hunk 6]
```diff
@@ -39,14 +39,5 @@ func (d Diagnostic) String() string {
...
 }
```
Comment (waynesun09) [inline internal/harness/lint.go:42]: **[medium] design** — Lint() method body is now empty -- consider adding a doc comment noting future intent

The Lint() method now creates `var diags []Diagnostic` and immediately returns it (which returns nil since no diagnostics are appended). While the Phase 4 plan explicitly says to keep Lint() for future lint rules, the current implementation allocates a slice variable that is never used. A minor style concern: `return nil` would be clearer than declaring an unused variable. More importantly, per ADR-0045 design decision D3, Lint() is intentionally separate from Validate() as a future extension point. The plan document (docs/plans/adr-0045-forge-portable-harness-phase4.md) mentions future rules like 'missing slug', 'single-forge informational', and 'stale base SHA'. Consider adding a brief comment in the method body noting these planned rules, so future contributors understand this is not dead code. Current implementation: `var diags []Diagnostic; return diags` -- suggest simplifying to `return nil` since the variable is unused.

File: internal/harness/lint_test.go [hunk 7]
```diff
@@ -7,21 +7,11 @@ import (
 )
 
 func TestLint(t *testing.T) {
-	t.Run("role set", func(t *testing.T) {
+	t.Run("valid harness returns nil", func(t *testing.T) {
```
Comment (waynesun09) [inline internal/harness/lint_test.go:10]: **[medium] testing** — Missing negative test for Lint() after role promotion -- the 'role empty' case is deleted without replacement

The base branch had a test `TestLint/role empty` that verified Lint() returned a warning when role was empty. The PR correctly removes this since Validate() now catches it. However, there is no new test verifying that Lint() returns nil (no diagnostics) when called on a valid harness with role set. The PR does add `t.Run("valid harness returns nil", ...)` which covers this, but there should also be a test confirming the boundary behavior: that Lint() does NOT produce a diagnostic for role when role IS present. The current tests only check the positive case. Consider adding a test that explicitly constructs a Harness with all fields valid and asserts `len(h.Lint()) == 0` to lock in that no spurious diagnostics leak through. This is a minor gap -- the existing 'valid harness returns nil' test partially covers it.

File: internal/harness/harness_test.go [hunk 8]
```diff
@@ -68,6 +69,7 @@ skills:
 func TestLoad_ValidationLoopMissingScript(t *testing.T) {
```
Comment (waynesun09) [inline internal/harness/harness_test.go:69]: **[medium] testing** — TestLoad_MissingAgent test no longer exercises role-missing path first

The test `TestLoad_MissingAgent` creates a YAML with only `skills:` (no agent, no role). Since Validate() now checks role before the existing pattern-validation block, the error ordering depends on which check runs first. Looking at the Validate() implementation, `agent` is checked before `role` (line ~288 checks agent=="", line ~298 checks role==""). So this test still works correctly -- it hits the agent error first. However, there is no explicit test for the case where agent IS present but role is missing. The test `TestLoad_RoleMissing` covers this (YAML has `agent: agents/test.md` without role), which is good. No code change needed, but worth noting that test coverage for the error ordering is implicitly rather than explicitly validated.

File: docs/plans/adr-0045-forge-portable-harness-phase4.md [hunk 9]
```diff
@@ -94,7 +94,7 @@ Every consumer of the removed code, and the action taken:
 ## PR Dependency Graph
 
 ```
-PR 1 (require role in Validate)  [independent]
+PR 1 (require role in Validate)  [independent]  ✅ Done (#2446)
```
Comment (waynesun09) [inline docs/plans/adr-0045-forge-portable-harness-phase4.md]: **[medium] design** — Plan marks PR 1 as Done while PR is still open and unmerged

The Phase 4 plan diff changes the PR 1 line to include a checkmark emoji and `#2446` reference, marking it as done. However, PR #2446 is still in `open` state and has not been merged. Marking a PR as done in the plan before it merges creates a discrepancy if the PR is subsequently revised or closed. Best practice is to update the plan document only after the PR merges, or use a different marker like 'In Review' rather than 'Done'. This is a minor process concern but could cause confusion if someone reads the plan and assumes the work has landed.

File: internal/cli/lock_test.go [hunk 10]
```diff
@@ -1198,8 +1207,8 @@ func TestRunLock_URLBaseAndURLRefsNoOrgConfig(t *testing.T) {
 	assert.Contains(t, err.Error(), "config.yaml")
 }
 
-func TestRunLock_LintWarningOnMissingRole(t *testing.T) {
-	// Verifies that runLock emits a lint warning when harness has no role.
+func TestRunLock_ErrorOnMissingRole(t *testing.T) {
```
Comment (waynesun09) [inline internal/cli/lock_test.go:1210]: **[medium] testing** — TestRunLock_ErrorOnMissingRole added but only tests the LoadWithBase path

The new test `TestRunLock_ErrorOnMissingRole` verifies that `runLock` fails when the harness has no role. This is a valuable addition. However, it creates the agent file on disk (agents/code.md) so ValidateFilesExist would pass, but the harness is loaded via `LoadWithBase` which calls `Validate()` internally. The test only covers the `runLock` code path. There is no corresponding test in `run_test.go` for `runAgent` failing when role is missing. The existing `TestRunAgent_HarnessLoadPipeline` and similar tests were updated to include `role: test`, but no negative test was added for the run command path. Consider adding a `TestRunAgent_ErrorOnMissingRole` test to verify that the `fullsend run` command also rejects harnesses without role, ensuring both major consumer paths are covered.

File: internal/harness/harness.go (hunk 1 above)
Comment (waynesun09) [inline internal/harness/harness.go:199]: **[medium] adr-alignment** — DiscoverAgents skips harnesses with empty role, but role is now required by Validate()

(Note: This finding relates to `internal/harness/discover.go:62` which is not in this diff, posted here on the related Validate() change.)

DiscoverAgents uses LoadRaw (which skips validation) and then filters out entries where both Role and Slug are empty (discover.go line 62: `if h.Role == "" && h.Slug == "" { continue }`). Now that role is required by Validate(), a harness file without role is invalid. However, DiscoverAgents silently skips it rather than reporting it as an error.

This creates a discoverability gap: if a user has an invalid harness file (missing role), DiscoverAgents will silently omit it from results without any indication that the file was skipped due to missing role. The user may not realize their harness is misconfigured until they try to run or lock it.

Suggestion: Since role is now required, consider changing DiscoverAgents to collect a warning diagnostic when a harness file has no role (similar to how parse errors are collected), or at minimum log a warning. This would help users discover misconfigured harness files during discovery rather than only at run time. Alternatively, since this is pre-existing behavior and the PR scope is narrowly about Validate(), this could be deferred to a follow-up.

**Review by waynesun09 — COMMENTED** [review]
ADR-0045 Review Squad — 11 findings (medium+ severity). Reviewed for correctness, security, and ADR alignment.

File: internal/harness/harness.go (hunk 1 above)
Comment (ggallen) [reply internal/harness/harness.go:199]: Fixed — restored omitempty on the Role tag. Validate() is the real enforcement mechanism, and this keeps consistency with Slug and Model.

File: internal/harness/harness.go (hunk 5 above)
Comment (ggallen) [reply internal/harness/harness.go:200]: Fixed — restored omitempty (same fix as the sibling comment). ADR struct example also reverted to match.

File: internal/harness/lint.go (hunk 2 above)
Comment (ggallen) [reply internal/harness/lint.go]: Fixed — simplified body to `return nil`. The existing doc comment above the function (lines 36-39) already explains its purpose as a post-Validate extension point.

File: internal/harness/lint.go (hunk 6 above)
Comment (ggallen) [reply internal/harness/lint.go:42]: Fixed — simplified to `return nil`. The existing doc comment on lines 36-39 already documents intent. Future lint rules (slug-missing, single-forge informational, stale base SHA) are tracked in the phase 4 plan.

File: internal/cli/run_test.go (hunk 3 above)
Comment (ggallen) [reply internal/cli/run_test.go:1820]: Fixed — tightened assertion to `"invalid harness: role field is required"` to confirm the error comes from the Load→Validate() path specifically.

File: internal/harness/harness_test.go (hunk 4 above)
Comment (ggallen) [reply internal/harness/harness_test.go:1115]: Acknowledged — agreed this is minor duplication. The Load()-level test exercises the full pipeline (file I/O + unmarshal + validate) which is still valuable as an integration test alongside the unit-level Validate() tests.

File: internal/harness/lint_test.go (hunk 7 above)
Comment (ggallen) [reply internal/harness/lint_test.go:10]: Acknowledged — the existing "valid harness returns nil" and "role and slug set" subtests cover this boundary. They confirm Lint() returns nil for a valid harness, which locks in that no spurious diagnostics leak.

File: internal/harness/harness_test.go (hunk 8 above)
Comment (ggallen) [reply internal/harness/harness_test.go:69]: Acknowledged — no code change needed. The error ordering is implicitly tested: TestLoad_MissingAgent hits the agent error first (checked before role in Validate()), and TestLoad_RoleMissing hits the role error when agent is present.

File: docs/plans/adr-0045-forge-portable-harness-phase4.md (hunk 9 above)
Comment (ggallen) [reply docs/plans/adr-0045-forge-portable-harness-phase4.md]: Fixed — changed status to "🔄 In Review (#2446)" to reflect the PR's actual state.

File: internal/cli/lock_test.go (hunk 10 above)
Comment (ggallen) [reply internal/cli/lock_test.go:1210]: TestRunAgent_ErrorOnMissingRole already exists at run_test.go:1799 — both consumer paths (runAgent and runLock) are covered. Also tightened both assertions to match the full error wrapping.

File: internal/harness/harness.go (hunk 1 above)
Comment (ggallen) [reply internal/harness/harness.go:199]: Acknowledged — agreed this is pre-existing behavior outside the scope of this PR. DiscoverAgents uses LoadRaw (which intentionally skips validation) for discovery purposes. Adding a warning there would be a good follow-up but belongs in a separate change.
